context.go

```diff
@@ -301,58 +311,18 @@ func (context *Context) TotalTimeouts() uint64 {
 
 // Stats returns the cumulative time spent in various parts of the WAF, all in nanoseconds
 // and the timeout value used
-func (context *Context) Stats() map[string]time.Duration {
-	return context.metrics.copy()
-}
-
-type metricsStore struct {
-	data  map[string]time.Duration
-	mutex sync.RWMutex
-}
-
-func (metrics *metricsStore) add(key string, duration time.Duration) {
-	metrics.mutex.Lock()
-	defer metrics.mutex.Unlock()
-	if metrics.data == nil {
-		metrics.data = make(map[string]time.Duration, 5)
-	}
-
-	metrics.data[key] += duration
-}
-
-func (metrics *metricsStore) get(key string) time.Duration {
-	metrics.mutex.RLock()
-	defer metrics.mutex.RUnlock()
-	return metrics.data[key]
-}
-
-func (metrics *metricsStore) copy() map[string]time.Duration {
-	metrics.mutex.Lock()
-	defer metrics.mutex.Unlock()
-	if metrics.data == nil {
-		return nil
-	}
-
-	copy := make(map[string]time.Duration, len(metrics.data))
-	for k, v := range metrics.data {
-		copy[k] = v
-	}
-	return copy
-}
+func (context *Context) Stats() Stats {
+	context.mutex.Lock()
+	defer context.mutex.Unlock()
 
-// merge merges the current metrics with new ones
-func (metrics *metricsStore) merge(other map[string]time.Duration) {
-	metrics.mutex.Lock()
-	defer metrics.mutex.Unlock()
-	if metrics.data == nil {
-		metrics.data = make(map[string]time.Duration, 5)
+	truncations := make(map[TruncationReason][]int, len(context.truncations))
+	for reason, counts := range context.truncations {
+		truncations[reason] = append([]int(nil), counts...)
```

Comment:
Smells like this could result in silly allocation patterns?

```suggestion
		truncations[reason] = make([]int, len(counts))
		copy(truncations[reason], counts)
```